clio-client/src/main/scala/org/broadinstitute/clio/client/dispatch/PatchExecutor.scala

```diff
@@ -0,0 +1,146 @@
+package org.broadinstitute.clio.client.dispatch
+
+import akka.NotUsed
+import akka.stream.scaladsl.Source
+import better.files.File
+import cats.syntax.either._
+import io.circe.Json
+import io.circe.syntax._
+import org.broadinstitute.clio.client.commands.PatchCommand
+import org.broadinstitute.clio.client.util.IoUtil
+import org.broadinstitute.clio.client.webclient.ClioWebClient
+import org.broadinstitute.clio.transfer.model.ClioIndex
+import org.broadinstitute.clio.JsonUtils.JsonOps
+
+class PatchExecutor[CI <: ClioIndex](patchCommand: PatchCommand[CI]) extends Executor {
+
+  import patchCommand.index.implicits._
+  import Executor.SourceMonadOps
+
+  private type DocumentKey = patchCommand.index.KeyType
+
+  private val queryAllFile: File =
+    File.newTemporaryFile().deleteOnExit().write(Json.obj().pretty(implicitly))
+
+  /**
+    * Build a stream which, when pulled, will communicate with the clio-server
+    * to update its records about some metadata, potentially performing IO
+    * operations in the process.
+    */
+  override def execute(
+    webClient: ClioWebClient,
+    ioUtil: IoUtil
+  ): Source[Json, NotUsed] = {
+    // Read the metadata then turn into Json so that we fail fast.
+    ioUtil
+      .readMetadata(patchCommand.index)(patchCommand.metadataLocation)
+      .map(_.asJson.dropNulls)
+      .flatMap(Source.repeat)
+      // zip new metadatas with documents
+      .zip(
+        webClient
+          .jsonFileQuery(patchCommand.index)(queryAllFile)
+          .map(_.dropNulls)
+      )
+      // Munge metadata for documents
+      .map {
+        case (newMetadataJson, documentJson) =>
+          mungeMetadata(newMetadataJson, documentJson)
+      }
+      // filter out things that already have no new metadata to upsert
+      .filterNot {
+        case (mergedJson, oldMetadataJson, _) =>
+          mergedJson.equals(oldMetadataJson)
+      }
+      // Upsert the stuff!
+      .flatMapConcat {
+        case (mergedMetadataJson, oldMetadataJson, docKey) =>
+          upsertPatch(mergedMetadataJson, oldMetadataJson, docKey, webClient)
+      }
+  }
+
+  /**
+    * Munge the patch metadata with the document metadata to get a new metadata json
+    * @param patchMetadataJson The new metadata to insert into the index
+    * @param documentJson The old document to insert the metadata into
+    * @return A tuple containing the munged metadata, the old metadata, and the key for this document
+    */
+  private def mungeMetadata(
+    patchMetadataJson: Json,
+    documentJson: Json
+  ): (Json, Json, DocumentKey) = {
+    val oldMetadataJson = getMetadataJsonFromDocument(documentJson)
+    val indexKey = getKeyFromDocument(documentJson)
+    (
+      patchMetadataJson.deepMerge(oldMetadataJson),
+      oldMetadataJson,
+      indexKey
+    )
+  }
+
+  /**
+    * Upsert the new metadata into the document.
+    * This method strips any existing keys out of the metadata to be upserted in order to reduce traffic.
```

Comment:
The term `keys` here is somewhat confusing is that it's not a clio index key, but a metadata field name, which happens to be a `key` in the map that represents the metadata.